typegate/src/typegate/artifacts/shared.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+
+import { createHash } from "node:crypto";
+import { S3, S3ClientConfig } from "aws-sdk/client-s3";
+import {
+  deinitUploadUrlStore,
+  getLocalPath,
+  initUploadUrlStore,
+  STORE_DIR,
+  STORE_TEMP_DIR,
+  UploadUrlStore,
+} from "./local.ts";
+import { ArtifactStore } from "./mod.ts";
+import { HashTransformStream } from "../../utils/hash.ts";
+import { resolve } from "std/path/resolve.ts";
+
+export class SharedArtifactStore extends ArtifactStore {
+  #uploadUrls: UploadUrlStore;
+  #s3!: S3;
+  #s3Bucket: string;
+
+  static init(s3Config: S3ClientConfig, s3Bucket: string) {
+    return new SharedArtifactStore(s3Config, s3Bucket);
+  }
+
+  constructor(s3Config: S3ClientConfig, s3Bucket: string) {
+    super();
+    this.#s3 = new S3(s3Config);
+    this.#s3Bucket = s3Bucket;
+    this.#uploadUrls = initUploadUrlStore();
+  }
+
+  override async persist(stream: ReadableStream<any>): Promise<string> {
+    const tmpFile = await Deno.makeTempFile({
+      dir: STORE_TEMP_DIR,
+    });
+    const file = await Deno.open(tmpFile, { write: true, truncate: true });
+    const hasher = createHash("sha256");
+    await stream
+      .pipeThrough(new HashTransformStream(hasher))
+      .pipeTo(file.writable);
+
+    const hash = hasher.digest("hex");
+    await this.#s3.putObject({
+      Bucket: this.#s3Bucket,
+      Key: hash,
```

Comment:
The s3 bucket would be shared with the typegraph store, so I suggest to add some prefix to the key for an organization purpose. Something like `/artifacts-cache/${hash}`...

Reply:
Ok, will do that.

Reply:
Done.